stats/stats.go

```diff
@@ -0,0 +1,140 @@
+package stats
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/elliotchance/orderedmap"
+	"github.com/hashicorp/go-hclog"
+	"github.com/segmentio/stats/v4"
+)
+
+type Measure struct {
+	id       string
+	duration time.Duration
+	stamp    bool
+	time     time.Time
+}
+
+type Stat struct {
+	Start    time.Time
+	Duration time.Duration
+}
+
+type LogHandler struct {
+	logger   hclog.Logger
+	measures chan Measure
+	stats    *orderedmap.OrderedMap
+	mu       sync.Mutex
+}
+
+type Options struct {
+	Logger  hclog.Logger
+	Tick    time.Duration
+	Handler stats.Handler
+}
+
+func NewClockWithObserve(name string, tags ...stats.Tag) *stats.Clock {
+	cl := stats.DefaultEngine.ClockAt(name, time.Now(), tags...)
+	stats.DefaultEngine.Observe(name, time.Duration(0), tags...)
+	return cl
+}
+
+func meta(name string, tags []stats.Tag) (string, bool) {
+	var stamp = false
+	var s []string
+	s = append(s, name)
+	for _, t := range tags {
+		// stamp is added on clock.Stop()
+		// we want that both clock.Start() and clock.Stop() have the same map id
+		if t.Name != "stamp" {
+			s = append(s, t.Name, t.Value)
+		} else {
+			stamp = true
+		}
+	}
+	return strings.Join(s, ":"), stamp
+}
+
+func (h *LogHandler) HandleMeasures(time time.Time, measures ...stats.Measure) {
+	for _, m := range measures {
+		id, stamp := meta(m.Fields[0].Name, m.Tags)
+		h.measures <- Measure{id: id, duration: m.Fields[0].Value.Duration(), time: time, stamp: stamp}
+	}
+}
+
+func (h *LogHandler) Flush() {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	hasItems := true
+	for hasItems {
+		select {
+		case m := <-h.measures:
+			if m.stamp {
+				item, ok := h.stats.Get(m.id)
+				if ok {
+					h.stats.Set(m.id, Stat{Start: item.(Stat).Start, Duration: m.duration})
+				}
+			} else {
+				h.stats.Set(m.id, Stat{Start: m.time, Duration: 0})
+			}
+		default:
+			hasItems = false
+		}
+	}
+
+	durationReported := make([]string, 0)
+	for el := h.stats.Front(); el != nil; el = el.Next() {
```

Comment:
why not use a queue structure rather then a map?

Reply:
We group 2 different measurements, `start` and `stop` by the same id, hence the usage of a map to access existing measurements.

A queue might not provide easy access to existing measurements.

The heartbeat keeps printing a running operation until a `clock.Stop` was called and then it will print the duration once.

Reply:
I am very confused fromcode as it only does a simple print every X this solution seems over complex. ill need to look again on the code and run it so I can give you better feedback.

Reply:
I'm happy to sync on this. It keeps track of the operations that were started and completed.
On every tick, it logs the tracked operations either with `running_for` if the operation wasn't completed or with `duration` if the operation was completed.
If the operation was completed is won't be printed again.